packages/lesswrong/lib/collections/posts/views.ts

```diff
@@ -823,24 +824,27 @@ ensureIndex(Posts, {legacyId: "hashed"});
 
 
 // Corresponds to the postCommented subquery in recentDiscussionFeed.ts
-ensureIndex(Posts,
-  {
-    status: 1,
-    isFuture: 1,
-    draft: 1,
-    unlisted: 1,
-    authorIsUnreviewed: 1,
-    hideFrontpageComments: 1,
-    
-    lastCommentedAt: -1,
-    _id: 1,
-    
-    baseScore: 1,
-    af: 1,
-    isEvent: 1,
-    globalEvent: 1,
-    commentCount: 1,
-  },
+const postCommentedViewFields = {
+  status: 1,
+  isFuture: 1,
+  draft: 1,
+  unlisted: 1,
+  authorIsUnreviewed: 1,
+  hideFrontpageComments: 1,
+  
+  lastCommentedAt: -1,
+  _id: 1,
+  
+  baseScore: 1,
+  af: 1,
+  isEvent: 1,
+  globalEvent: 1,
+  commentCount: 1,
+}
+ensureIndex(Posts, isEAForum ? {
```

Comment:
Dumb question: do you have to explicitly drop the old index somehow?

Reply:
I've done that manually in the past. I guess I'm not sure if I should include it in a migration now - @oetherington thoughts?

Reply:
You would need to drop the old index because this will end up having the same name (this is definitely something that could/should be improved).

This index isn't actually being used though. The relevant selector clause gets compiled into `("tagRelevance"->'ZCihBFp5P64JCvQY6')::INTEGER < 1 OR ("tagRelevance"->'ZCihBFp5P64JCvQY6') IS NULL` which can't make use of a GIN index (which is what we use for any JSON fields), and fixing it is actually non-trivial.

I managed to make some decent progress by manually casting `"tagRelevance"->'ZCihBFp5P64JCvQY6'` to an integer when creating the index and changing the clause to `("tagRelevance" ? 'ZCihBFp5P64JCvQY6') AND ("tagRelevance"->'ZCihBFp5P64JCvQY6')::INTEGER < 1` but it still ends up doing a heap scan afterwards for the boolean fields.

That's already a factor of 10 better than it was before, but we could still do better by eliminating that last heap scan. I'll try to dig deeper in a bit.

Reply:
Ok, I've experimented a bit more and I think there are four things we should do to fix this:
 - Instead of `"tagRelevance"->'ZCihBFp5P64JCvQY6' IS NOT NULL` the query builder should emit `NOT ("tagRelevance" ? 'ZCihBFp5P64JCvQY6')` for JSON objects
 - Remove this special index - I did manage to find a version that can work, but it's pointless because we only perform the table scan _after_ we `LIMIT 10`, so adding a special index actually hurts performance
 - In order to match mongo behaviour, whenever we output an `ORDER BY` we use `DESC NULLS LAST` and `ASC NULLS FIRST` instead of just `DESC` and `ASC` which prevents us from using indexes in some cases (such as here). We either need to add explicit ordering to the indexes or add an option to not emit the null ordering specification. Here we probably just want to do the latter since the `WHERE` clause prevents us from getting any NULLs here anyway
 - In the `recentDiscussionsFeed` resolver we `$and` together the "events criteria" and "exclude community" selectors which creates another level of nesting preventing Postgres from realising that it can use the index - we should add these separately (#6618 adds a `mergeSelectors` function which would be really useful here)

I'll try to do all of this later today.

Reply:
@oetherington Thanks so much!

> https://github.com/ForumMagnum/ForumMagnum/pull/6618 adds a mergeSelectors function which would be really useful here

Isn't that function just doing what I did in this PR, or am I missing something?

Reply:
Oh yeah - you're right, I got confused. We only need to do 1, 2 and 3.